### PR TITLE
AP-9818 # added `generateMorphExpressionPropertyHandler` to `submissi…

### PR DESCRIPTION
**Pull request by divporter**
…onService`

## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "fixpack": "^4.0.0",
     "globals": "^17.0.0",
     "lodash": "^4.17.23",
+    "morph-expressions": "^1.1.1",
```
Comment (mymattcarroll) [inline package.json:27]: Making this a dev dependency just means we have to maintain it in 3 code bases going forward. Much easier to maintain long term if this is a dependency.

The value of having no dependencies in this code base is about reliably being able to run the code on a browser or server without the risk of the dependency running browser or nodejs code specifically (e.g. accessing global variables like `window`). By using this morph-expressions library (indirectly) in sdk-core, we are losing that guarantee already. No need to strive for no deps if we are relying on a another lib during runtime anyway.

Could abstract the `morph-expressions` away from calling code all together. Proposal:

```typescript
function evaluateCalculationExpression(options: {
    expression: string;
    submission: SubmissionTypes.S3SubmissionData['submission'];
    formElements: FormTypes.FormElement[];
    parseDayOnlyDate: ParseDayOnlyDate;
}): number | undefined;
```

Comment (mymattcarroll) [reply package.json:27]: I (cursor) had a crack at my idea: https://github.com/oneblink/sdk-core-js/pull/150

Comment (divporter) [reply package.json:27]: We run date-fns code in sdk-core all the time from our own codebases. We lost the below a long time ago.
> reliably being able to run the code on a browser or server without the risk of the dependency running browser > or nodejs code specifically 

Is this compatible with apps-react land?

Comment (mymattcarroll) [reply package.json:27]: > We run date-fns code in sdk-core all the time from our own codebases

`sdk-core` executes callbacks which may or may not run `date-fns` code (server side we use `date-fns`, but typically client side we just use built in browser functions). The idea is to let the calling code decide which external libs to execute.

This PR would mean that sdk-core will always execute code from the `morph-expressions` lib (or at least a class that has the same interface as `morph-expressions` lib). I believe we either want to:

1. Have a callback to do the actually transformation from "expression" to `number` (like we do for date parsing so that the calling code can do the timezone magic)

or

2. Have the calling code not worry about the `morph-expressions` lib at all and have sdk-core maintain that code.

There is not really any value in maintaining two ways of evaluating our expression (compared to parsing dates which has value server vs client side). In this PR, we are using the `morph-expressions` lib both client and server side and if we decide to stop using it, we would very likely stop using it both sides as well.

Option 2 seems the most appropriate but if you would prefer to go with this PR, im on board 👍 .

Comment (divporter) [reply package.json:27]: @kizaonline is the story leader. I don't really mind what we proceed with as long as this plays nice in apps-react which has a lot of stuff broken up in various memos and whatnot.

Comment (mymattcarroll) [reply package.json:27]: I believe my proposed PR will play nice in react land.

Comment (divporter) [reply package.json:27]: Closing in favour of https://github.com/oneblink/sdk-core-js/pull/150
